File: solutions/java/maximum-subarray.java
// id: 006
// title: Maximum Subarray
// approach: Kadane's Algorithm
// time: O(n) | space: O(1)
// author: chron303

import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine().trim());
        String[] input = br.readLine().trim().split(" ");

        int[] nums = new int[n];
        for (int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(input[i]);
        }

        int currentSum = nums[0];
        int maxSum = nums[0];

        for (int i = 1; i < n; i++) {
            currentSum = Math.max(nums[i], currentSum + nums[i]);
            maxSum = Math.max(maxSum, currentSum);
        }

        System.out.println(maxSum);
    }
}

File: solutions/python/maximum-subarray.py
# id: 006
# title: Maximum Subarray
# approach: Kadane's Algorithm
# time: O(n) | space: O(1)
# author: chron303

import sys

def main():
    data = sys.stdin.read().split()
    n = int(data[0])
    nums = list(map(int, data[1:n+1]))

    current_sum = nums[0]
    max_sum = nums[0]

    for i in range(1, n):
        current_sum = max(nums[i], current_sum + nums[i])
        max_sum = max(max_sum, current_sum)

    print(max_sum)

if __name__ == "__main__":
    main()
